node/src/components/contract_runtime/operations.rs

```diff
@@ -155,61 +131,121 @@ pub fn execute_finalized_block(
-
-    let execution_effect: AdditiveMap<Key, Transform> = match ee_execution_result {
-        EngineExecutionResult::Success {
-            execution_journal,
-            cost,
-            ..
-        } => {
-            // We do want to see the deploy hash and cost in the logs.
-            // We don't need to see the effects in the logs.
-            debug!(?deploy_hash, %cost, "execution succeeded");
-            execution_journal
+) -> Result<(ExecResultMap, Digest), BlockExecutionError> {
+    let block_time = finalized_block.timestamp().millis();
+
+    // Create a new EngineState that reads from LMDB but only caches changes in memory.
+    let scratch_state = engine_state.get_scratch_engine_state();
+
+    let deploy_items = deploys
+        .iter()
+        .cloned()
+        .map(DeployItem::from)
+        .collect::<Vec<_>>();
+
+    let mut execution_results = BTreeMap::new();
+
+    for deploy_item in deploy_items {
+        // Note that we generate a request to execute all deploys in the block.
+        let execute_request = ExecuteRequest::new(
+            state_root_hash,
+            block_time,
+            vec![deploy_item],
+            protocol_version,
+            finalized_block.proposer(),
+        );
+
+        // Execute our request against the in-memory cache.
+        let Execution { exec_results } = execute(&scratch_state, metrics.clone(), execute_request)?;
+
+        let transforms_by_deploy = extract_transforms_from_execution_results(&exec_results);
+
+        // This ensures that transforms are applied in the order they were executed.
+        for (_deploy_hash, transforms) in transforms_by_deploy.into_iter() {
+            // Because we are applying the transforms to the ScratchGlobalState, no new
+            // state-root-hash is generated here.
+            let _same_root_hash =
+                apply_transforms(&scratch_state, metrics.clone(), state_root_hash, transforms)?;
         }
-        EngineExecutionResult::Failure {
-            error,
-            execution_journal,
-            cost,
-            ..
-        } => {
-            // Failure to execute a contract is a user error, not a system error.
-            // We do want to see the deploy hash, error, and cost in the logs.
-            // We don't need to see the effects in the logs.
-            debug!(?deploy_hash, ?error, %cost, "execution failure");
-            execution_journal
+
+        for (deploy_hash, ee_exec_result) in exec_results {
+            let node_deploy_hash: NodeDeployHash = deploy_hash.into();
+            if let Some(deploy) = deploys
+                .iter()
+                .find(|deploy| node_deploy_hash == *deploy.id())
+            {
+                execution_results.insert(
+                    node_deploy_hash,
+                    (
+                        deploy.header().clone(),
+                        ExecutionResult::from(&ee_exec_result),
+                    ),
+                );
+            }
```

Comment:
Did you actually mean to pass in all `deploy_items` to `ExecRequest`? If not, all of these iterations over the results are unnecessary as there will always be just one `execution result`.